Feat : #113

Young-Yeong opened the pull request:
close : #113

Young-Yeong committed:
Feat : 키오스크 관련 Api 개발

- Key
    - DynamicUpdate 적용

- KeyRepository
    - KeyRepositoryCustom 상속
    - findByRfid
        - rfid로 차키 조회
    - findByKeyStorage
        - 차량 보관함으로 차키 조회

- KeyStorage
    - DynamicUpdate 적용

- KeyStorageRepository
    - KeyStorageRepositoryCustom 상속

- KeyStorageRepositoryCustom
    - findKeyStorageByReservationId
        - 예약 번호로 차 키 보관함 조회
    - findUsableSlotByKioskId
        - 키오스크의 사용가능한 슬롯 조회

- KeyStorageRepositoryImpl
    - findKeyStorageByReservationId
        - 입력받은 예약번호로 차량 조회하고 나서 차량으로 차 키를 조회하고 그 후 차 키로 차량보관함 조회
    - findUsableSlotByKioskId
        - 요청한 키오스크에 남은 슬롯 조회

- KioskController
    - getKeyStorageByReservationId
        - GET /api/v1/branches/kiosks/keyStorage
            - request : String name
                        Long reservationId
                        Long kioskId
            - response : Long
        - POST /api/va/branches/kiosk/return
            - request : Map<String,String> rfid
                        Long kioskId
            - response : Long

- KioskService
    - findKeyStorage
        - 입력받은 이름과 예약ID로 조회 후 유효성 검사 후 차 키 위치 반환
    - returnKey
        - 입력받은 rfid로 유효성 검사 후 사용 가능한 자리 조회 후 반환

- ReservationRepositoryCustom
    - findRecentReservationsByCarId
        - 차량에 대한 최근 예약 조회

- ReservationRepositoryImpl
    - findRecentReservationsByCarId
        - 입력받은 carId에 해당하는 최근 예약 조회

Related to : #113

Capoomaru commented on src/main/java/com/yu/yurentcar/domain/branch/controller/KioskController.java:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    @GetMapping("/keyStorage")
+    public ResponseEntity<Long> getKeyStorageByReservationId(@RequestParam String name, @RequestParam Long reservationId, @RequestParam Long kioskId) {
```
`requestParam` 을 묶어서 하나의 dto로 받아오는게 좋아보여요 :)

Capoomaru commented on src/main/java/com/yu/yurentcar/domain/branch/controller/KioskController.java:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    @PostMapping("/return")
+    public ResponseEntity<Long> returnKey(@RequestBody Map<String, String> rfid, @RequestParam Long kioskId) {
```
왜 Map으로 받아오나요? Map으로 받아오는게 제대로 작동할 지 모르겠네요 + kioskId 도 body에 넣어서 한 번에 다받아도 될 것 같아요

Young-Yeong replied:
로컬 테스트 결과 제대로 작동합니다.

Capoomaru replied:
좋습니다! 근데 후자는 반영해주세요 :)

Capoomaru commented on src/main/java/com/yu/yurentcar/domain/branch/entity/KeyStorage.java:
```diff
@@ -24,10 +26,16 @@ public class KeyStorage extends BaseTimeEntity {
+    @Column(name = "is_available")
     private Boolean isAvailable;
+
+    public KeyStorage updateAvailable(Boolean isAvailable) {
```
👍👍

Capoomaru commented on src/main/java/com/yu/yurentcar/domain/branch/service/KioskService.java:
```diff
@@ -0,0 +1,128 @@
+            throw new RuntimeException("차 키가 존재하지 않습니다. 관리자에게 문의하세요");
+        }
+
+       /* // 파이썬 서버로 요청 보내는 로직
+        RestTemplate restTemplate = new RestTemplate();
+        // 헤더 설정
+        HttpHeaders headers = new HttpHeaders();
+        headers.setContentType(MediaType.APPLICATION_JSON);
+
+        log.info("kioskId.toString() = " + kioskId.toString());
+        // 파라미터 세팅
+        MultiValueMap<String, String> map = new LinkedMultiValueMap<>();
+        map.add("kioskId", kioskId.toString());
+        map.add("slotNumber", keyStorage.getKeyStorageId().toString());
+
+        HttpEntity<MultiValueMap<String, String>> request = new HttpEntity<>(map, headers);
+        String url = "http://localhost:8080/api/v1/test/keyStorage";
+
+        Boolean response = restTemplate.postForObject(url, request, Boolean.class);
+        if (Boolean.FALSE.equals(response)) {
+            throw new RuntimeException("열리지 않았습니다. 다시 시도해주세요");
+        }*/
+
```
제대로 동작하는 진 모르겠지만 RestTemplate 잘 적용했네요 :)

Capoomaru reviewed (approved):
LGTM 👍

Young-Yeong committed:
Fix : 키오스크 관련 Api 수정

- KeyReturnDto
    - 차 키 반납 Dto

- KioskController
    - returnKey
        - 파라미터 수정
Related to : #114
